cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTableAdmin.java

```diff
@@ -141,13 +174,56 @@ private boolean tableExistsInternal(Connection connection, StructuredTableId tab
     }
   }
 
+  // Add new non-primary key columns to table
+  private void addColumns(Connection connection, StructuredTableSchema schema)
+    throws IOException, TableNotFoundException {
+    StructuredTableId tableId = schema.getTableId();
+    try {
+      Optional<String> columnStatement = getAddColumnStatement(tableId,
+                                                               connection.getMetaData(), schema);
+      if (!columnStatement.isPresent()) {
+        LOG.debug("No new columns to add for table {}", schema.getTableId());
+        return;
+      }
+      try (Statement statement = connection.createStatement()) {
+        LOG.debug("Create adding columns statement: {}", columnStatement);
+        statement.execute(columnStatement.get());
+      }
+    } catch (SQLException e) {
+      // PostgreSQL error 42P01 denotes the database query is on an undefined table.
+      if (e.getClass().getName().equalsIgnoreCase("org.postgresql.util.PSQLException")
+        && "42P01".equalsIgnoreCase(e.getSQLState())) {
+        throw new TableNotFoundException(tableId);
+      }
+      throw new IOException(String.format("Error adding columns to table: %s", tableId.getName()), e);
+    }
+  }
+
+  // Add new non-primary key indices to table
+  private void addIndices(Connection connection, StructuredTableSchema schema)
+    throws IOException, TableNotFoundException {
+    try (Statement statement = connection.createStatement()) {
+      // Create indexes
+      for (String indexStatement : getCreateIndexStatements(schema.getTableId(),
+                                                            getNonExistIndexes(connection, schema))) {
+        LOG.debug("Create index statement: {}", indexStatement);
+        statement.execute(indexStatement);
+      }
+    } catch (SQLException e) {
+      if (e.getClass().getName().equalsIgnoreCase("org.postgresql.util.PSQLException")
```

Comment:
Do we need this class name check or is just checking the `getSQLState` is sufficient?

Reply:
great qn, this `42P01` error code seems to be only used in PSQLException. Think the getSQLState is sufficient.